src/node/node_state.h

```diff
@@ -374,42 +375,74 @@ namespace ccf
-            launch_node();
+            // On SEV-SNP, if no attestation report endorsements are set via
+            // environment, those need to be fetched
+            quote_endorsements_client =
+              std::make_shared<QuoteEndorsementsClient>(
+                rpcsessions,
+                endpoint_config,
+                [this, qi](std::vector<uint8_t>&& endorsements) {
+                  std::lock_guard<pal::Mutex> guard(lock);
+                  quote_info = qi;
+                  quote_info.endorsements = std::move(endorsements);
+                  try
+                  {
+                    launch_node();
+                  }
+                  catch (const std::exception& e)
+                  {
+                    LOG_FAIL_FMT("{}", e.what());
+                    throw;
+                  }
+                  quote_endorsements_client.reset();
+                });
+
+            quote_endorsements_client->fetch_endorsements();
             return;
           }
 
-          quote_endorsements_client = std::make_shared<QuoteEndorsementsClient>(
-            rpcsessions,
-            endpoint_config,
-            [this, quote_info_](std::vector<uint8_t>&& endorsements) {
-              // Note: Only called for SEV-SNP
-              std::lock_guard<pal::Mutex> guard(lock);
-              quote_info = quote_info_;
-              quote_info.endorsements = std::move(endorsements);
-              try
-              {
-                launch_node();
-              }
-              catch (const std::exception& e)
-              {
-                LOG_FAIL_FMT("{}", e.what());
-                throw;
-              }
-              quote_endorsements_client.reset();
-            });
+          CCF_ASSERT_FMT(
+            (qi.format == QuoteFormat::oe_sgx_v1 && !qi.endorsements.empty()) ||
+              (qi.format != QuoteFormat::oe_sgx_v1 && qi.endorsements.empty()),
+            "SGX quote generation should have already fetched endorsements");
 
-          quote_endorsements_client->fetch_endorsements();
+          quote_info = qi;
+
+          if (
+            quote_info.format == QuoteFormat::amd_sev_snp_v1 &&
+            config.attestation.environment.report_endorsements.has_value())
+          {
+            // On SEV-SNP, if reports endorsements are passed via
+            // environment, read those rather than fetching them from
+            // endorsement server
+            pal::snp::ACIReportEndorsements endorsements =
+              nlohmann::json::parse(crypto::raw_from_b64(
+                config.attestation.environment.report_endorsements.value()));
+
+            CCF_ASSERT_FMT(
+              quote_info.endorsements.empty(),
+              "No endorsements should be set by quote generation");
+
+            quote_info.endorsements.insert(
+              quote_info.endorsements.end(),
+              endorsements.vcek_cert.begin(),
+              endorsements.vcek_cert.end());
```

Comment:
Is it inserting `std::string` ( = `std::basic_string<char>`) to `std::vector<uint8_t>` ?
If so it might be unsafe because char might not be 8 bit. 
https://en.cppreference.com/w/cpp/language/types
(Practically it should be fine because this code runs only on SEV-SNP machine)

Reply:
That's a good point. The CCF codebase has a few assumptions about the underlying platform (e.g. little endian) as we're only targeting x86. In this case, the endorsements are a PEM-encoded string so that should be OK?